app_grim_reefer/src/main.c

```diff
@@ -138,7 +136,68 @@ static int sensor_init(void) {
     return 0;
 }
 
-#define EVENT_FILTER_ALL 0xFFFFFFFF
+// Flight Phase
+K_TIMER_DEFINE(flight_duration_timer, NULL, NULL);
+K_TIMER_DEFINE(camera_extra_timer, NULL, NULL);
+// State Machine Setup
+enum flight_state { PAD_STATE, FLIGHT_STATE, LANDED_STATE };
+const struct smf_state flight_states[];
+struct {
+    struct smf_ctx ctx;
+    /* All User Defined Data Follows */
+} s_obj;
+
+// On Pad, boost detecting
+static void pad_state_entry(void *o) {
+    LOG_INF("On Pad");
+    start_boost_detect(lsm6dsl_dev, bme280_dev);
+}
+static void pad_state_run(void *o) {
+    if (get_boost_detected()) {
+        smf_set_state(SMF_CTX(&s_obj), &flight_states[FLIGHT_STATE]);
+    }
+}
+static void pad_state_exit(void *o) { stop_boost_detect(); }
+
+// In flight, data sampling
+static void flight_state_entry(void *o) {
+    LOG_INF("Flight Started");
+    // Turn on Cameras and ADC
+    gpio_pin_set_dt(&ldo_enable, 1);
+    gpio_pin_set_dt(&cam_enable, 1);
+    k_timer_start(&flight_duration_timer, TOTAL_FLIGHT_TIME, K_NO_WAIT);
+}
+static void flight_state_run(void *o) {
+
+    if (k_timer_status_get(&flight_duration_timer) > 0) {
+        // timer has expired
+        smf_set_state(SMF_CTX(&s_obj), &flight_states[LANDED_STATE]);
+    }
+}
+static void flight_state_exit(void *o) {}
+
+// Landed, saving files and waiting for cameras
+static void landed_state_entry(void *o) {
+    // Stop logging, start saving
+    LOG_INF("Landed");
+    k_timer_start(&camera_extra_timer, CAMERA_EXTRA_TIME, K_NO_WAIT);
+
+    finish_data_storage();
+    buzzer_tell(buzzer_cond_landed);
+    save_boost_data();
+}
+static void landed_state_run(void *o) {
+    if (k_timer_status_get(&camera_extra_timer) > 0) {
+        // timer has expired
+        smf_set_terminate(SMF_CTX(&s_obj), 1);
+    }
```

Comment:
k_timer_status_sync instead of a spinlock?

Reply:
I was under the impression that you don't want your `run` function to block or else main will never get control back to run `smf_run_state`. I suppose in this case it wouldn't matter either way since this is the only way out of the state but it felt like good practice. Dyk about this? Will research / test when I get a sec.

Reply:
Good point. This might make sense, but I would test. If your main isn't really doing anything else it should be fine